### PR TITLE
Initial route prefix mini-RFC proposal

### PR DESCRIPTION
**Pull request by AlexMSmithCA**
# Summary

A route prefix plugin for FusionJS.

### CODE REVIEW

File: text/0000-route-prefix.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  onRoute={...}
+>{child}</Router>
+```
+[(link)](https://github.com/fusionjs/fusion-plugin-react-router#router)
+
+As user land plugins should not be accessing environment variables directly, this will likely be supplied by a user land `{config}.js` file.  This file must remain synchronized to the configuration used in the static asset plugin and the ctx re-write.  Due to this being sensitive to errors, it may be prudent to have a build time check here to ensure synchronization.
+
+# Drawbacks
+
+This is a non-trival amount of effort to implement.  In its current form, it is also prone to user error and requires duplicated configuration.  As noted below, the alternative in the [dependency injection](https://github.com/fusionjs/rfcs/pull/1/commits/8c440259291c9774b44b9dca07cf35dbf2a5f98e?short_path=9919d4f#diff-9919d4f45f05666e94a1bcb5edd7f51a) world is potentially cleaner.  Implementing the proposal above will require a migration path towards the DI implementation down the road.
```
Comment (KevinGrandon) [inline text/0000-route-prefix.md]: Can you expand on which files would be duplicates? The yaml file with the basepath definition and what else? I probably missed the other place.

Comment (AlexMSmithCA) [reply text/0000-route-prefix.md]: Sure!  At a high level, the two places would likely be `pinocchio.yaml` and a user land config file, like `routePrefixConfig.js`.

* The yaml file would power the environment variables used by `fusion-cli` which includes populating [webpack's public path](https://github.com/fusionjs/fusion-cli/search?utf8=%E2%9C%93&q=__webpack_public_path__&type=) and [`assetUrl`](https://github.com/fusionjs/fusion-cli/blob/af9da1ff34be8c9546f6de5c75b87850024d1e1a/build/babel-plugins/babel-plugin-asseturl/index.js).
* The user land configuration would be used to configure user land plugins - which should not access environment variables - like [`fusion-plugin-react-router`](https://github.com/fusionjs/fusion-plugin-react-router#router) and [`fusion-plugin-font-loader-react`](https://github.com/fusionjs/fusion-plugin-font-loader-react/search?utf8=%E2%9C%93&q=_static&type=).

File: text/0000-route-prefix.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+# Alternatives
+
+Alternatively, the open proposal for migrating towards a purer [dependency injection](https://github.com/fusionjs/rfcs/pull/1/commits/8c440259291c9774b44b9dca07cf35dbf2a5f98e?short_path=9919d4f#diff-9919d4f45f05666e94a1bcb5edd7f51a) system for the framework may yield a cleaner solution here as well.  At a high level, the solution may involve the following concerns:
+
+* Creating a `RoutePrefixPluginConfigToken` token to use for dependency injection.
```
Comment (AlexMSmithCA) [inline text/0000-route-prefix.md]: For what it's worth, the naming here could be better... more than likely this configuration is a single string that other plugins also depend upon.  Something like `routePrefixStringToken` may be a more accurate name.

File: text/0000-route-prefix.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    // enhance ctx.url sans prefix
+    if(ctx.url.indexOf(prefix) === 0 /*found at index 0*/) {
+      ctx.url = ctx.url.slice(prefix.length);
+    }
+  }
+}
+```
+
+### Prefix handling for `fusion-plugin-react-router`
```
Comment (AlexMSmithCA) [inline text/0000-route-prefix.md]: @mlmorg noted that there may exist other user-land plugins that will require the route prefix as a configuration property.  We'll want to identify all of these are part of this process.

**Pull request closed without merging**